### PR TITLE
BFD-3604: Document Claim and ClaimResponse POST Endpoints

### PR DESCRIPTION
**Pull request by MahiFentaye**
**JIRA Ticket:**
BFD-3604
BFD-3605

### What Does This PR Do?
Adds swagger documentation for POST endpoints Claim and ClaimResponse

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

### What Security Implications Does This PR Have?

Please indicate if this PR does any of the following:  

* Adds any new software dependencies
* Modifies any security controls
* Adds new transmission or storage of data
* Any other changes that could possibly affect security? 

* [x] I have considered the above security implications as it relates to this PR. (If one or more of the above apply, it cannot be merged without the ISSO or team security engineer's (`@sb-benohe`) approval.) 

### Validation

Have you fully verified and tested these changes? Is the acceptance criteria met? Please provide reproducible testing instructions, code snippets, or screenshots as applicable.

### CODE REVIEW

**Review by aschey-forpeople — APPROVED** [review]

**Review by dondevun — APPROVED** [review]

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml [hunk 1]
```diff
@@ -354,3 +354,241 @@ paths:
...
+                        - `includeTaxNumbers=true`
+                    example: true
+                  isHashed:
+                    type: string
+                    description: |-
+                      Setting this flag to _true_, provides tax number in the EOB transformed data for the response.
+
+                      Example:
+                        - `isHashed=true`
+                              example: "ABCD"
```
Comment (mjburling) [inline apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml:427]: This has both some formatting and content challenges... I don't think the content-related issues are limited to this specific block, either. Hmm.

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml [hunk 2]
```diff
@@ -354,3 +354,241 @@ paths:
...
+                      - `includeTaxNumbers=true`
+                  example: true
+                isHashed:
+                  type: string
+                  description: |-
+                    Setting this flag to _true_, provides tax number in the EOB transformed data for the response.
+                    
+                    Example:
+                      - `isHashed=true`
+                            example: "ABCD"
```
Comment (mjburling) [inline apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml:546]: Noted elsewhere, this needs to be reviewed for accuracy.

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml [hunk 3]
```diff
@@ -354,3 +354,241 @@ paths:
...
+                  example: "fiss,mcs"
+        required: true
+      responses:
+        '200':
+          description: Success
+          content:
+            application/fhir+json:
+              schema: { $ref: '#/components/schemas/FHIR-JSON-RESOURCE' }
+            application/fhir+xml:
+              schema: { $ref: '#/components/schemas/FHIR-XML-RESOURCE' }
```
Comment (mjburling) [inline apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml:594]: There's _a very_ minor issue with the no newline at EOF warning here. Could you make adjustments to your editor, write/save, and add a new commit to address this?

https://thoughtbot.com/blog/no-newline-at-end-of-file#following-the-rules-in-your-editor

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml [hunk 4]
```diff
@@ -354,3 +354,241 @@ paths:
...
+
+                      ```
+                      A combination of upper case ASCII letters ('A'..'Z except for S, L, O, I, B, and Z.', numerals ('0'..'9'),
+                      '-', with a length limit of 11 characters.
+                      ```
+
+                      The following are all valid values for MBI, and all might
+                      represent the same resource:
+                        - `mbi=9AB2WW3GR44` (unhashed MBI)
+                        - `mbi=82273caf4d2c3b5a8340190ae3575950957ce469e593efd7736d60c3b39d253c` (hashed)
```
Comment (mjburling) [inline apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml:385]: I haven't looked _too_ closely at the implementation, but the spirit of enabling (and documenting) the POST endpoints is, in part, to discourage the use of this hashed identifier under RFC20.

* Are hashed MBIs honored here?
* If they are accepted, would it be better to omit them from this documentation _or_ somehow label them as deprecated?

**Review by mjburling — CHANGES_REQUESTED** [review]
Thanks for the submission, Mahi! I'm really looking forward to having the POST methods documented for Claim and ClaimResponse.

Overall, this is shaping up! However, there are a few formatting and content issues that need to be reviewed and addressed before this can be accepted. In addition to a few in-line comments... have we considered correcting the content type to represent the what I believe is the required `application/x-www-form-urlencoded` as opposed to `application/json`?

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml (hunk 1 above)
Comment (MahiFentaye) [reply apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml:427]: I will work on the formatting, I will updating the existing doc as well because they all follow the same formatting. The content is copied from the recent `OpenAPI.yaml`, should I have a different source?

Comment (mjburling) [reply apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml:427]: Right–it must be inaccurate in the source material you're referring to. I gather `isHashed` is a temporary input that's used with an `mbi` input to signal whether the input should be treated as hashed or plaintext, e.g. `mbi=1234&isHashed=false`.

Again, I admit that I haven't looked too closely at the implementation... can you also document what the default behavior is? If `isHashed` is omitted, does the API treat the inputs as hashed or unhashed by default?

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml [hunk 5]
```diff
@@ -354,3 +354,241 @@ paths:
...
+                        - `excludeSAMHSA=true`
+                    example: true
+                  identifier:
+                    type: string
+                    includeTaxNumbers: |-
+                      Setting this flag to _true_, provides tax number in the EOB transformed data for the response.
+
+                      Example:
+                        - `includeTaxNumbers=true`
+                    example: true
```
Comment (mjburling) [inline apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml:419]: Having another look at this... `identifier` and/or `includeTaxNumbers` don't look quite right.

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml (hunk 1 above)
Comment (MahiFentaye) [reply apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml:427]: looks like `isHashed` is required. without specifying `isHashed=false ` It doesn't return anything but with returns all the values expected.

```
{
  "resourceType": "Bundle",
  "id": "12345",
  "meta": {
    "lastUpdated": "2024-09-10"
  },
  "type": "searchset",
  "total": 0,
  "link": [
    {
      "relation": "self",
      "url": "https://localhost:6500/v2/fhir/Claim?mbi=12345"
    }
  ]
}
```

Comment (MahiFentaye) [reply apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml:427]: default is as setting `isHashed=true`

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml [hunk 6]
```diff
@@ -350,7 +345,235 @@ paths:
...
+                  description: |-
+                    When fetching a _Bundle Response_ using pagination, this URL parameter represents an offset
+                    (starting point) into the list of elements for the _Request_.
+                    It is optional and defaults to 1 if not supplied.
+                    A value 0 is not allowed and negative indices are not currently supported.
+                    
+                    Example:
+                    - `startIndex=100`
+                  example: 100
+                type:
```
Comment (alex-dzeda) [inline apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml:558]: why is this required?

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml [hunk 7]
```diff
@@ -350,7 +345,235 @@ paths:
...
+        requestBody:
+          content:
+            application/x-www-form-urlencoded:
+              schema:
+                type: object
+                properties:
+                  mbi:
+                    type: string
+                    description: |-
+                      **NOTE: THIS IS A REQUIRED FIELD**
```
Comment (alex-dzeda) [inline apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml:364]: If this is required, it should have that in the "required" syntax rather than a bolded note.

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml [hunk 8]
```diff
@@ -350,7 +345,235 @@ paths:
...
+
+                      Supported Claim Type values:
+                        - `fiss`
+                        - `mcs`
+
+                      Examples:
+                        - `type=fiss,mcs`
+                        - `type=fiss`
+                    example: "fiss,mcs"
+          required: true
```
Comment (alex-dzeda) [inline apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml:458]: Why is type required?

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml [hunk 9]
```diff
@@ -122,18 +121,16 @@ paths:
         '200':
           description: Success
           content:
-            application/fhir+json:
-              schema: {$ref: '#/components/schemas/FHIR-JSON-RESOURCE'}
-            application/fhir+xml:
-              schema: {$ref: '#/components/schemas/FHIR-XML-RESOURCE'}
+            application/x-www-form-urlencoded:
+              schema:
```
Comment (mjburling) [inline apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml:125]: These code the content of the response rather than the request–I don't think these need to be adjusted here or elsewhere under the content sections.

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml [hunk 10]
```diff
@@ -350,7 +345,235 @@ paths:
...
+      tags: [v2,ClaimResponse]
+      summary: 'search-type: Search for ClaimResponse instances'
+      description: This is a search type
+      requestBody:
+        content:
+          application/x-www-form-urlencoded:
+            schema:
+              type: object
+              properties:
+                mbi:
```
Comment (alex-dzeda) [inline apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml:475]: "required" property is appropriate instead of putting it in the description

**Review by alex-dzeda — CHANGES_REQUESTED** [review]
EOB != claim (EOB is a finalized claim), should be updated.
I am not sure about the use of "required" in documentation vs. required, and type appears to be required?

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml [hunk 11]
```diff
@@ -61,7 +61,6 @@ paths:
                     
                     Example:
                        - `includeTaxNumbers=true`
-                  enum: ["true", "false"]
```
Comment (mjburling) [inline apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml:64]: I'm not clear on this... did this need to be removed?

Comment (MahiFentaye) [reply apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml:64]: I ended up removing this to make it consistent with the others

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml (hunk 4 above)
Comment (MahiFentaye) [reply apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml]: removed hashed mbis

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml (hunk 5 above)
Comment (MahiFentaye) [reply apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml]: fixed

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml (hunk 9 above)
Comment (MahiFentaye) [reply apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml]: fixed

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml [hunk 12]
```diff
@@ -351,6 +349,234 @@ paths:
...
+                    represents the medicare benficiary ID, or as much of that
+                    defined by the [FHIR specification](https://terminology.hl7.org/NamingSystem-cmsMBI.html) as:
+                    
+                    ```
+                    A combination of upper case ASCII letters ('A'..'Z except for S, L, O, I, B, and Z.', numerals ('0'..'9'),
+                    '-', with a length limit of 11 characters.
+                    ```
+                    
+                    The following are all valid values for MBI, and all might
+                    represent the same resource:
```
Comment (dondevun) [inline apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml:494]: I'm not sure `all might represent the same resource` holds true here, since you removed the hashed MBI

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml (hunk 4 above)
Comment (MahiFentaye) [reply apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml]: Hashed mbis still work on POST with the `isHashed` flag `true`

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml [hunk 13]
```diff
@@ -351,6 +349,232 @@ paths:
...
+                    ```
+                    
+                    The following are all valid values for MBI:
+                      - `mbi=1EG4TE5MK73` (unhashed MBI)
+                      - `mbi=1S00E00HA26` (synthetic MBI)
+                  example: "1EG4TE5MK73"
+                _lastUpdated:
+                  type: string
+                  description: |-
+                    Only satisfy the Search if the Beneficiary's `last_updated` Date falls within a specified _DateRange_.
```
Comment (mjburling) [inline apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml:499]: Here (and elsewhere), there's an issue with this text, but I'm not sure if it shouldn't be more generalized. It's easy to miss that this is specific to _this_ FHIR resource... maybe it shouldn't read "Claim's" rather, "this FHIR resource's"?

```suggestion
                    Only satisfy the Search if the Claim's `last_updated` Date falls within a specified _DateRange_.
```

**Review by dondevun — DISMISSED** [review]

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml [hunk 14]
```diff
@@ -351,6 +349,232 @@ paths:
...
+                      
+                      Example:
+                        - `isHashed=true`
+                  service-date:
+                    type: string
+                    description: |-
+                      Only satisfy the Search request if a claim's Date
+                      falls within a specified _DateRange_. A _DateRange_ can be
+                      defined by providing less than `lt` and/or greater than `gt` values.
+                      This parameter can be included in a request one or more times.
```
Comment (mjburling) [inline apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml:428]: Do we know _which_ date this is referring to? Does it apply to just a subset, e.g. FISS vs MCS?

**Review by mjburling — CHANGES_REQUESTED** [review]
This is shaping up! Had a few additional comments on this round of reviews for you to consider. Thanks!

File: apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml (hunk 14 above)
Comment (MahiFentaye) [reply apps/bfd-server/bfd-server-war/src/test/resources/openapi/posts.yaml]: description updated to : The _DateRange_ is defined by the claim's billing period, specifically the `Claim.billablePeriod.end`.

**Review by dondevun — APPROVED** [review]

**Review by aschey-forpeople — APPROVED** [review]

**Review by mjburling — APPROVED** [review]
